### PR TITLE
Table of contents for CollectionsPage

### PR DESCRIPTION
**Pull request by Raemon**
Adds a ToC, and the ability to manually specify book titles

image

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1203496669044111) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/books/schema.ts [hunk 1]
```diff
@@ -29,6 +29,16 @@ const schema: SchemaType<DbBook> = {
...
   },
 
+  // this overrides the book title in the CollectionsPage table of contents,
+  // for books whose title needs to be different there for whatever reason.
+  tocTitle: {
+    type: String,
+    optional: true,
+    viewableBy: ['guests'],
+    editableBy: ['members'],
+    insertableBy: ['members'],
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/books/schema.ts:39]: nit: prefer `can[Action]` fields

File: packages/lesswrong/lib/collections/books/schema.ts [hunk 2]
```diff
@@ -29,6 +29,16 @@ const schema: SchemaType<DbBook> = {
     insertableBy: ['members'],
   },
 
+  // this overrides the book title in the CollectionsPage table of contents,
+  // for books whose title needs to be different there for whatever reason.
+  tocTitle: {
+    type: String,
+    optional: true,
```
Comment (b0b3rt) [inline packages/lesswrong/lib/collections/books/schema.ts:36]: should also be `nullable`?

File: packages/lesswrong/components/sequences/CollectionTableOfContents.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    book.sequences.forEach(sequence => {
+      if (sequence.title) {
+        sections.push(({
+          title: sequence.title,
+          anchor: sequence._id,
+          level: 2
+        }))
+      }
+    })
+  })
```
[the remark this answers is not in the document]
Comment (Raemon) [reply packages/lesswrong/components/sequences/CollectionTableOfContents.tsx:51]: In this case it actually would instead be a bug to use `anchor: slugify(book.tocTitle || book.title)`. The anchor needs to match the id given to the SectionTitle elsewhere on the page, which is always based on book.title.

I do agree flatMap seems nicer, I forgot that that option existed (I'd briefly tried a regular map and it failed and then I gave up on that pattern)

**Review by b0b3rt — APPROVED** [review]
lgtm, just make `tocTitle` nullable and leave a comment disambiguating the title logic in this bit (if not refactoring to pull out variables):
```
    if (book.tocTitle || book.title) {
      sections.push(({
        title: book.tocTitle || book.title,
        anchor: slugify(book.title),
```

**Pull request merged**